### DISCUSSION TITLE
How to specify the `default_prog` with command substitution?

### DISCUSSION THREAD

**Discussion by ylorn** [wezterm/wezterm#2868] [Q&A]
Say I have 2 Macs, one uses Intel chip, and the other uses Apple Silicon. Both of them install `fish` via `HomeBrew`, and therefore the `fish` is installed at different locations on 2 machines. But I want them to share the same config file. So how can I achieve something like:

```lua
  default_prog = { '$(brew --prefix)/bin/fish', '-l' },
```

**Comment by wez** [comment]
You have a couple of options:

# 1. Set the path up in the wezterm process, so it can see all your homebrew utilities

This works around the heart of the problem, which is that the PATH isn't set up for homebrew by macOS,
by setting up the PATH explicitly.

```lua
local wezterm = require 'wezterm'

function brew_prefix()
  local success, stdout, stderr = wezterm.run_child_process { 'brew', '--prefix' }
  if success then
    -- strip trailing newline off `brew --prefix` output
    return stdout:gsub("%s+", "")
  end
end

function brew_paths()
  local prefix = brew_prefix()
  if prefix then
     return { prefix .. "/bin", prefix .. "/sbin" }
  end
  return {}
end

function compute_path()
  local path = os.getenv("PATH")
  for _, dir in ipairs(brew_paths()) do
      path = path .. ":" .. dir
  end
  return path
end

return {
  set_environment_variables = {
      PATH = compute_path(),
  },
  default_prog = {'fish', '-l'}
}
```

See also:
* https://wezfurlong.org/wezterm/config/lua/config/set_environment_variables.html
* https://wezfurlong.org/wezterm/config/lua/wezterm/run_child_process.html

# 2. Just use the prefix for the default program

Similar to the above, but is focused on just getting a working `default_prog`:

```lua
local wezterm = require 'wezterm'

function brew_prefix()
  local success, stdout, stderr = wezterm.run_child_process { 'brew', '--prefix' }
  if success then
    -- strip trailing newline off `brew --prefix` output
    return stdout:gsub("%s+", "")
  end
end

return {
  default_prog = { brew_prefix() .. "/bin/fish", "-l" }
}
```

# 3. Use the shell

This is much more succinct, but is a bit less efficient, as it requires spawning both the shell and the brew utility for each new pane.
In practice, that won't matter.

```lua
return {
    default_prog = {"sh", "-c", "exec $(brew --prefix)/bin/fish -l"},
}
```

**Reply by ylorn** [reply]
Lovely, thanks for the swift reply!

**Reply by ylorn** [reply]
Since the path of `$(brew --prefix)` is already in my environment variable `PATH` for bash/zsh/fish on my system, so I tried with:

```lua
return {
  set_environment_variables = {
      PATH = os.getenv('PATH'),
  },
  -- Spawn a fish shell in login mode
  default_prog = { 'fish', '-l' },
}
```

but the WezTerm still crashed on launch.

**Reply by wez** [reply]
wezterm is typically launched via Finder, and neither of those are launched by your shell and thus won't see a custom PATH environment.  So you either need to replicate your path setup in your wezterm config, or have wezterm explicitly launch things via the shell.

**Reply by wez** [reply]
also: wezterm doesn't crash on launch in the case of a bad path, it just exits because the program you launched exited.  You can set 
`exit_behavior = 'CloseOnCleanExit'` to see more information in that case.

https://wezfurlong.org/wezterm/config/lua/config/exit_behavior.html

**Reply by ylorn** [reply]
> `exit_behavior = 'CloseOnCleanExit'` to see more information in that case.

Very useful. Here is what I got.

For the previous config, 

```lua
return {
  set_environment_variables = {
      PATH = os.getenv('PATH'),
  },
  exit_behavior = 'CloseOnCleanExit',
  -- Spawn a fish shell in login mode
  default_prog = { 'fish', '-l' },
}
```

The `WezTerm` does not hold with messages, but quit on launch. 

However, I tried another approach you suggested with following configuration:
```lua
return {
    exit_behavior = 'CloseOnCleanExit',
    default_prog = {"sh", "-c", "exec $(brew --prefix)/bin/fish -l"},
}
```

This time the `WezTerm` does not quit on launch, but got held with messages:

```
⚠️  Process "sh -c 'exec $(brew --prefix)/bin/fish -l'" in domain "local" didn't exit cleanly
Exited with code 126.
This message is shown because exit_behavior="CloseOnCleanExit"
```

**Reply by ylorn** [reply]
I tried with the 1st approach also:

```lua
function brew_prefix()
  local success, stdout, stderr = wezterm.run_child_process { 'brew', '--prefix' }
  if success then
    -- strip trailing newline off `brew --prefix` output
    return stdout:gsub("%s+", "")
  end
end

function brew_paths()
  local prefix = brew_prefix()
  if prefix then
     return { prefix .. "/bin", prefix .. "/sbin" }
  end
  return {}
end

function compute_path()
  local path = os.getenv("PATH")
  for _, dir in ipairs(brew_paths()) do
      path = path .. ":" .. dir
  end
  return path
end

return {
  set_environment_variables = {
      PATH = compute_path(),
  },
  default_prog = {'fish', '-l'}
  exit_behavior = 'CloseOnCleanExit',
}
```

but the config seems corrupted:
<img width="686" alt="Screenshot 2022-12-18 at 12 53 15 AM" src="https://user-images.githubusercontent.com/1502810/208283785-dadddabe-c769-42ec-8c05-58b7ee99ea1c.png">

**Reply by wez** [reply]
Sounds like `brew` is probably not in PATH either; I'd suggest trying option 4 below.

**Comment by wez** [comment]
# 4. Hardcode based on the architecture

If you don't want to spawn `brew` in the config file, or via the shell, you could do:

```lua
local wezterm = require 'wezterm'

local fish = 'fish'

if wezterm.target_triple == 'aarch64-apple-darwin'
  -- this is where brew installs on my M1 mac
  fish = "/opt/homebrew/bin/fish"
elseif wezterm.target_triple == 'x86_64-apple-darwin'
  fish = "/usr/local/bin/fish" -- or wherever it is on your intel mac
end

return {
  default_prog = {fish, "-l"}
}
```

See also:
* https://wezfurlong.org/wezterm/config/lua/wezterm/target_triple.html
